### PR TITLE
allow covariance to be formula

### LINKED ISSUE #291: covariance object not able to convert to cov_terms when using tidymodels
**Summary**

when trying to run mmrm through tidymodels framework passing the new covariance structure in the set_engine step it passes the right informaiton back to mmrm, but mmrm has an issue parsing back out the covar object into a formula and breaks [here](https://github.com/openpharma/mmrm/blob/main/R/utils-formula.R#L65)

``` r
library(mmrm)
library(tidymodels)
#> Warning: package 'tidymodels' was built under R version 4.2.3
#> 
#> Warning: package 'broom' was built under R version 4.2.3
#> Warning: package 'dials' was built under R version 4.2.3
#> Warning: package 'scales' was built under R version 4.2.2
#> Warning: package 'dplyr' was built under R version 4.2.3
#> Warning: package 'ggplot2' was built under R version 4.2.3
#> Warning: package 'infer' was built under R version 4.2.3
#> Warning: package 'modeldata' was built under R version 4.2.3
#> Warning: package 'parsnip' was built under R version 4.2.3
#> Warning: package 'purrr' was built under R version 4.2.2
#> Warning: package 'recipes' was built under R version 4.2.3
#> Warning: package 'rsample' was built under R version 4.2.3
#> Warning: package 'tibble' was built under R version 4.2.3
#> Warning: package 'tidyr' was built under R version 4.2.2
#> Warning: package 'tune' was built under R version 4.2.3
#> Warning: package 'workflows' was built under R version 4.2.3
#> Warning: package 'workflowsets' was built under R version 4.2.3
#> Warning: package 'yardstick' was built under R version 4.2.3
linear_reg() |>
  set_engine("mmrm", control = mmrm_control(method = 'Satterthwaite'), covariance = ~us(AVISIT | USUBJID)) |>
  fit(FEV1 ~ RACE + ARMCD * AVISIT, fev_data)
#> Error in eval(substitute(expr), data, enclos = parent.frame()): invalid 'envir' argument of type 'language'
details::details(sessioninfo::session_info(), summary = 'si')
```

<details closed>
<summary>
<span title="Click to Expand"> si </span>
</summary>

``` r

─ Session info ───────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.2.1 (2022-06-23 ucrt)
 os       Windows 10 x64 (build 19044)
 system   x86_64, mingw32
 ui       RTerm
 language (EN)
 collate  English_United States.utf8
 ctype    English_United States.utf8
 tz       America/New_York
 date     2023-06-14
 pandoc   3.1.1 @ C:/Program Files/RStudio/resources/app/bin/quarto/bin/tools/ (via rmarkdown)

─ Packages ───────────────────────────────────────────────────────────────────
 package      * version    date (UTC) lib source
 backports      1.4.1      2021-12-13 [1] CRAN (R 4.2.0)
 broom        * 1.0.4      2023-03-11 [1] CRAN (R 4.2.3)
 checkmate      2.2.0      2023-04-27 [1] CRAN (R 4.2.3)
 class          7.3-20     2022-01-16 [3] CRAN (R 4.2.1)
 cli            3.6.1      2023-03-23 [1] CRAN (R 4.2.3)
 clipr          0.8.0      2022-02-22 [1] CRAN (R 4.2.1)
 codetools      0.2-18     2020-11-04 [3] CRAN (R 4.2.1)
 colorspace     2.1-0      2023-01-23 [1] CRAN (R 4.2.2)
 data.table     1.14.2     2021-09-27 [1] CRAN (R 4.2.1)
 desc           1.4.2      2022-09-08 [1] CRAN (R 4.2.1)
 details        0.3.0      2022-03-27 [1] CRAN (R 4.2.2)
 dials        * 1.2.0      2023-04-03 [1] CRAN (R 4.2.3)
 DiceDesign     1.9        2021-02-13 [1] CRAN (R 4.2.3)
 digest         0.6.31     2022-12-11 [1] CRAN (R 4.2.2)
 dplyr        * 1.1.2      2023-04-20 [1] CRAN (R 4.2.3)
 evaluate       0.21       2023-05-05 [1] CRAN (R 4.2.3)
 fansi          1.0.4      2023-01-22 [1] CRAN (R 4.2.2)
 fastmap        1.1.0      2021-01-25 [1] CRAN (R 4.2.1)
 foreach        1.5.2      2022-02-02 [1] CRAN (R 4.2.3)
 fs             1.6.2      2023-04-25 [1] CRAN (R 4.2.3)
 furrr          0.3.1      2022-08-15 [1] CRAN (R 4.2.3)
 future         1.32.0     2023-03-07 [1] CRAN (R 4.2.3)
 future.apply   1.11.0     2023-05-21 [1] CRAN (R 4.2.3)
 generics       0.1.3      2022-07-05 [1] CRAN (R 4.2.3)
 ggplot2      * 3.4.2      2023-04-03 [1] CRAN (R 4.2.3)
 globals        0.16.2     2022-11-21 [1] CRAN (R 4.2.2)
 glue           1.6.2      2022-02-24 [1] CRAN (R 4.2.2)
 gower          1.0.1      2022-12-22 [1] CRAN (R 4.2.2)
 GPfit          1.0-8      2019-02-08 [1] CRAN (R 4.2.3)
 gtable         0.3.2      2023-03-17 [1] CRAN (R 4.2.1)
 hardhat        1.3.0      2023-03-30 [1] CRAN (R 4.2.3)
 highr          0.9        2021-04-16 [1] CRAN (R 4.2.1)
 htmltools      0.5.4      2022-12-07 [1] CRAN (R 4.2.2)
 httr           1.4.4      2022-08-17 [1] CRAN (R 4.2.1)
 infer        * 1.0.4      2022-12-02 [1] CRAN (R 4.2.3)
 ipred          0.9-14     2023-03-09 [1] CRAN (R 4.2.3)
 iterators      1.0.14     2022-02-05 [1] CRAN (R 4.2.3)
 knitr          1.41       2022-11-18 [1] CRAN (R 4.2.2)
 lattice        0.21-8     2023-04-05 [1] CRAN (R 4.2.3)
 lava           1.7.2.1    2023-02-27 [1] CRAN (R 4.2.3)
 lhs            1.1.6      2022-12-17 [1] CRAN (R 4.2.3)
 lifecycle      1.0.3      2022-10-07 [1] CRAN (R 4.2.2)
 listenv        0.9.0      2022-12-16 [1] CRAN (R 4.2.3)
 lubridate      1.9.2      2023-02-10 [1] CRAN (R 4.2.3)
 magrittr       2.0.3      2022-03-30 [1] CRAN (R 4.2.2)
 MASS           7.3-58.1   2022-08-03 [1] CRAN (R 4.2.2)
 Matrix         1.5-4.1    2023-05-18 [1] CRAN (R 4.2.3)
 mmrm         * 0.2.2.9017 2023-06-01 [1] local
 modeldata    * 1.1.0      2023-01-25 [1] CRAN (R 4.2.3)
 munsell        0.5.0      2018-06-12 [1] CRAN (R 4.2.2)
 nlme           3.1-162    2023-01-31 [1] CRAN (R 4.2.3)
 nnet           7.3-17     2022-01-16 [3] CRAN (R 4.2.1)
 numDeriv       2016.8-1.1 2019-06-06 [1] CRAN (R 4.2.0)
 parallelly     1.35.0     2023-03-23 [1] CRAN (R 4.2.3)
 parsnip      * 1.1.0      2023-04-12 [1] CRAN (R 4.2.3)
 pillar         1.9.0      2023-03-22 [1] CRAN (R 4.2.3)
 pkgconfig      2.0.3      2019-09-22 [1] CRAN (R 4.2.2)
 png            0.1-8      2022-11-29 [1] CRAN (R 4.2.2)
 prodlim        2023.03.31 2023-04-02 [1] CRAN (R 4.2.3)
 purrr        * 1.0.1      2023-01-10 [1] CRAN (R 4.2.2)
 R.cache        0.16.0     2022-07-21 [1] RSPM (R 4.2.0)
 R.methodsS3    1.8.2      2022-06-13 [1] CRAN (R 4.2.0)
 R.oo           1.25.0     2022-06-12 [1] CRAN (R 4.2.0)
 R.utils        2.12.2     2022-11-11 [1] CRAN (R 4.2.3)
 R6             2.5.1      2021-08-19 [1] CRAN (R 4.2.2)
 rbibutils      2.2.13     2023-01-13 [1] CRAN (R 4.2.3)
 Rcpp           1.0.10     2023-01-22 [1] CRAN (R 4.2.2)
 Rdpack         2.4        2022-07-20 [1] CRAN (R 4.2.1)
 recipes      * 1.0.6      2023-04-25 [1] CRAN (R 4.2.3)
 reprex         2.0.2      2022-08-17 [1] CRAN (R 4.2.2)
 rlang          1.1.1      2023-04-28 [1] CRAN (R 4.2.3)
 rmarkdown      2.19       2022-12-15 [1] CRAN (R 4.2.2)
 rpart          4.1.16     2022-01-24 [3] CRAN (R 4.2.1)
 rprojroot      2.0.3      2022-04-02 [1] CRAN (R 4.2.1)
 rsample      * 1.1.1      2022-12-07 [1] CRAN (R 4.2.3)
 rstudioapi     0.14       2022-08-22 [1] CRAN (R 4.2.2)
 scales       * 1.2.1      2022-08-20 [1] CRAN (R 4.2.2)
 sessioninfo    1.2.2      2021-12-06 [1] CRAN (R 4.2.2)
 stringi        1.7.12     2023-01-11 [1] CRAN (R 4.2.2)
 stringr        1.5.0      2022-12-02 [1] CRAN (R 4.2.2)
 styler         1.9.1      2023-03-04 [1] CRAN (R 4.2.3)
 survival       3.3-1      2022-03-03 [3] CRAN (R 4.2.1)
 tibble       * 3.2.1      2023-03-20 [1] CRAN (R 4.2.3)
 tidymodels   * 1.1.0      2023-05-01 [1] CRAN (R 4.2.3)
 tidyr        * 1.3.0      2023-01-24 [1] CRAN (R 4.2.2)
 tidyselect     1.2.0      2022-10-10 [1] CRAN (R 4.2.2)
 timechange     0.2.0      2023-01-11 [1] CRAN (R 4.2.3)
 timeDate       4022.108   2023-01-07 [1] CRAN (R 4.2.3)
 tune         * 1.1.1      2023-04-11 [1] CRAN (R 4.2.3)
 utf8           1.2.3      2023-01-31 [1] CRAN (R 4.2.2)
 vctrs          0.6.2      2023-04-19 [1] CRAN (R 4.2.3)
 withr          2.5.0      2022-03-03 [1] CRAN (R 4.2.2)
 workflows    * 1.1.3      2023-02-22 [1] CRAN (R 4.2.3)
 workflowsets * 1.0.1      2023-04-06 [1] CRAN (R 4.2.3)
 xfun           0.35       2022-11-16 [1] CRAN (R 4.2.2)
 xml2           1.3.3      2021-11-30 [1] CRAN (R 4.2.1)
 yaml           2.3.6      2022-10-18 [1] CRAN (R 4.2.2)
 yardstick    * 1.2.0      2023-04-21 [1] CRAN (R 4.2.3)

 [1] C:/Users/jonathan.sidi/AppData/Local/R/win-library/4.2
 [2] C:/Users/jonathan.sidi/AppData/Local/Programs/R/R-4.2.1/site-library
 [3] C:/Users/jonathan.sidi/AppData/Local/Programs/R/R-4.2.1/library

──────────────────────────────────────────────────────────────────────────────
```

</details>

<br>

<sup>Created on 2023-06-14 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

### PR DESCRIPTION
**Pull request by clarkliming**
close #291

### CODE REVIEW

File: R/cov_struct.R [hunk 1]
```diff
@@ -272,6 +272,11 @@ cov_struct <- function(
 #'
 #' @keywords internal
 h_reconcile_cov_struct <- function(formula = NULL, covariance = NULL) {
+  assert_multi_class(covariance, c("formula", "cov_struct"), null.ok = TRUE)
+  assert_formula(formula, null.ok = FALSE)
```
Comment (clarkliming) [inline R/cov_struct.R:276]: if set to null, also need to update `h_extract_covariance_terms` to allow it to work with NULL (and there should not be cases using this function with formula as NULL)

File: tests/testthat/test-cov-struct.R [hunk 2]
```diff
@@ -132,3 +132,21 @@ test_that("print.cov_struct pretty-formats covariance structure object", {
     "`VIS IT`, VISIT | GROUP / SUBJECT"
   )
 })
+
```
Comment (danielinteractive) [inline tests/testthat/test-cov-struct.R:135]: please also add an integration test that resembles Yoni's original problem, showing that it is fixed now. thanks

Comment (clarkliming) [reply tests/testthat/test-cov-struct.R:135]: added now

**Review by danielinteractive — COMMENTED** [review]
sorry unrelated but too small for new issue: please use `stats::na_omit` in `h_factor_ref` to avoid note in R CMD check

**Comment by clarkliming** [comment]
updated now

**Review by danielinteractive — APPROVED** [review]

**Pull request merged**